plutus-contract/src/Plutus/Contract/StateMachine.hs

```diff
@@ -135,13 +135,13 @@ getStates
     :: forall s i
     . (PlutusTx.FromData s, PlutusTx.ToData s)
     => SM.StateMachineInstance s i
-    -> Map Tx.TxOutRef Tx.ChainIndexTxOut
+    -> Map Tx.TxOutRef Tx.OffChainTxOut
     -> [OnChainState s i]
 getStates (SM.StateMachineInstance _ si) refMap =
-    flip mapMaybe (Map.toList refMap) $ \(txOutRef, ciTxOut) -> do
-      let txOut = Tx.toTxOut ciTxOut
-      datum <- ciTxOut ^? Tx.ciTxOutScriptDatum . _Right
-      ocsTxOutRef <- either (const Nothing) Just $ Typed.typeScriptTxOutRef si txOutRef txOut datum
+    flip mapMaybe (Map.toList refMap) $ \(txOutRef, ocTxOut) -> do
+      -- FIXME is it important to check txOut is a script?
+      datum <- Tx.ocTxOutDatum ocTxOut
+      ocsTxOutRef <- either (const Nothing) Just $ Typed.typeScriptTxOutRef si txOutRef (Tx.toTxOut ocTxOut) datum
```

Comment:
This funtion used `ciTxOutScriptDatum` which would only match when the txOut is protected by a script. Is this important? `ocTxOutDatum` on the other hand gives you a `Maybe Datum` independently from whether the txOut is protected by a pubkey or a script.

Reply:
I checked, I'm pretty sure this function only gets to see script outputs.

Reply:
Ok, so we can keep the `ocTxOutDatum` since at this point the difference is inessential right?

Reply:
Yes